gql_ws_link/lib/src/socket_client.dart

```diff
@@ -0,0 +1,372 @@
+      _socket.add(
+        json.encode(
+          message,
+          toEncodable: (dynamic m) => m.toJson(),
+        ),
+      );
+    }
+  }
+
+  /// Sends a query, mutation or subscription request to the server, and returns a stream of the response.
+  ///
+  /// If the request is a query or mutation, a timeout will be applied to the request as specified by
+  /// [SocketClientConfig]'s [queryAndMutationTimeout] field.
+  ///
+  /// If the request is a subscription, obviously no timeout is applied.
+  ///
+  /// In case of socket disconnection, the returned stream will be closed.
+  Stream<SubscriptionData> subscribe(
+      final QueryPayload payload, {@required final bool waitForConnection, @required final bool isSubscription }) {
+    final String id = Uuid.randomUuid(random: randomBytesForUuid).toString();
+    final StreamController<SubscriptionData> response =
+    StreamController<SubscriptionData>();
+    StreamSubscription<SocketConnectionState> sub;
+    final bool addTimeout = !isSubscription &&
+        config.queryAndMutationTimeout != null;
+
+    final onListen = () {
+      final Stream<SocketConnectionState>
+      waitForConnectedStateWithoutTimeout = _connectionStateController
+          .startWith(
+          waitForConnection ? null : SocketConnectionState.connected)
+          .where((SocketConnectionState state) =>
+      state == SocketConnectionState.connected)
+          .take(1);
+
+      final Stream<SocketConnectionState> waitForConnectedState = addTimeout
+          ? waitForConnectedStateWithoutTimeout.timeout(
+        config.queryAndMutationTimeout,
+        onTimeout: (EventSink<SocketConnectionState> event) {
+          print("Connection timed out.");
+          response.addError(TimeoutException("Connection timed out."));
+          event.close();
+          response.close();
+        },
+      )
+          : waitForConnectedStateWithoutTimeout;
+
+      sub = waitForConnectedState.listen((_) {
+        final Stream<GraphQLSocketMessage> dataErrorComplete =
+        _messageStream.where(
+              (GraphQLSocketMessage message) {
+            if (message is SubscriptionData) {
+              return message.id == id;
+            }
+
+            if (message is SubscriptionError) {
+              return message.id == id;
+            }
+
+            if (message is SubscriptionComplete) {
+              return message.id == id;
+            }
+
+            return false;
+          },
+        ).takeWhile((_) => !response.isClosed);
+
+        final Stream<GraphQLSocketMessage> subscriptionComplete = addTimeout
+            ? dataErrorComplete
+            .where((GraphQLSocketMessage message) =>
+        message is SubscriptionComplete)
+            .take(1)
+            .timeout(
+          config.queryAndMutationTimeout,
+          onTimeout: (EventSink<GraphQLSocketMessage> event) {
+            print("Request timed out.");
```

Comment:
The `print` statements should be dropped. People who want to debug should opt-in using the debugger.